File: src/test/java/edu/ucdavis/fiehnlab/ctsRest/controllers/CtsControllerTest.java
package edu.ucdavis.fiehnlab.ctsRest.controllers;

import edu.ucdavis.fiehnlab.ctsRest.ApplicationTests;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.cloud.netflix.feign.EnableFeignClients;
import org.springframework.http.MediaType;
import org.springframework.http.converter.HttpMessageConverter;
import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
import org.springframework.mock.http.MockHttpOutputMessage;
import org.springframework.mock.web.MockMultipartHttpServletRequest;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;
import org.springframework.web.context.WebApplicationContext;

import java.io.Console;
import java.io.IOException;

import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.hasSize;
import static org.junit.Assert.assertNotNull;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

/**
 * Created by diego on 2/15/2017.
 */
@RunWith(SpringRunner.class)
@SpringBootTest(value = "${fiehnlab.cts.config.name}")
@EnableFeignClients
public class CtsControllerTest extends ApplicationTests {
	private HttpMessageConverter mappingJackson2HttpMessageConverter;

	@Autowired
	private WebApplicationContext webApplicationContext;

	@Autowired
	void setConverters(HttpMessageConverter<?>[] converters) {

		for (HttpMessageConverter<?> hmc : converters) {
			if (hmc instanceof MappingJackson2HttpMessageConverter) {
				this.mappingJackson2HttpMessageConverter = hmc;
				break;
			}
		}

		assertNotNull("the JSON message converter must not be null", this.mappingJackson2HttpMessageConverter);
	}

	private MockMvc mockMvc;

	@Before
	public void setUp() {
		mockMvc = MockMvcBuilders.webAppContextSetup(webApplicationContext).build();
	}

	@Test
	public void getIndex() throws Exception {
		mockMvc.perform(get("/").accept(MediaType.APPLICATION_JSON))
				.andExpect(status().isOk())
				.andExpect(content().string(equalTo("CTS main page")));
	}

	@Test
	public void getSimpleConversion() throws Exception {
		MvcResult res = mockMvc.perform(MockMvcRequestBuilders.get("/rest/convert/chemical name/inchikey/ethanol").accept(MediaType.APPLICATION_JSON))
				.andExpect(status().isOk())
				.andExpect(jsonPath("$[0].fromIdentifier").value("chemical name"))
				.andExpect(jsonPath("$[0].toIdentifier").value("inchikey"))
				.andExpect(jsonPath("$[0].searchTerm").value("ethanol"))
				.andExpect(jsonPath("$[0].result").value("LFQSCWFLJHTTHZ-UHFFFAOYSA-N")).andReturn();

		System.out.println(res.getResponse().getContentAsString());
	}

	@Test
	public void getFormulaExpansion() throws Exception {
		MvcResult mvcres = mockMvc.perform(MockMvcRequestBuilders.get("/rest/expandformula/C2H6O").accept(MediaType.APPLICATION_JSON))
				.andExpect(status().isOk())
				.andExpect(jsonPath("$.result").value("CCHHHHHHO")).andReturn();

		System.out.println(mvcres.getResponse().getContentAsString());
	}

	@Test
	public void testGetFromValues() throws Exception {
		MvcResult mvcres = mockMvc.perform(MockMvcRequestBuilders.get("/rest/fromValues").accept(MediaType.APPLICATION_JSON))
				.andExpect(status().isOk())
				.andExpect(jsonPath("$", hasSize(224))).andReturn();

		System.out.println(mvcres.getResponse().getContentAsString());
	}

	@Test
	public void testGetToValues() throws Exception {
		MvcResult mvcres = mockMvc.perform(MockMvcRequestBuilders.get("/rest/toValues").accept(MediaType.APPLICATION_JSON))
				.andExpect(status().isOk())
				.andExpect(jsonPath("$", hasSize(225))).andReturn();

		System.out.println(mvcres.getResponse().getContentAsString());
	}

	@Test
	@Ignore("MockMvc doesn't have the X-Proxy-Cache header for this test to pass. Working fine in production")
	public void testSecondCallHitsCache() throws Exception {
		MvcResult res1 = mockMvc.perform(MockMvcRequestBuilders.get("/rest/convert/chemical name/inchikey/alanine").accept(MediaType.APPLICATION_JSON))
				.andExpect(status().isOk())
				.andReturn();
		Assert.assertEquals("MISS", res1.getResponse().getHeaderValue("X-Proxy-Cache"));

		MvcResult res2 = mockMvc.perform(MockMvcRequestBuilders.get("/rest/convert/chemical name/inchikey/alanine").accept(MediaType.APPLICATION_JSON))
				.andExpect(status().isOk())
				.andReturn();
		Assert.assertEquals("HIT", res2.getResponse().getHeaderValue("X-Proxy-Cache"));

	}


	private String json(Object o) throws IOException {
		MockHttpOutputMessage mockHttpOutputMessage = new MockHttpOutputMessage();
		this.mappingJackson2HttpMessageConverter.write(o, MediaType.APPLICATION_JSON, mockHttpOutputMessage);
		return mockHttpOutputMessage.getBodyAsString();
	}
}

File: web/src/test/scala/edu/ucdavis/fiehnlab/ctsrest/web/controllers/CtsControllerTest.scala
package edu.ucdavis.fiehnlab.ctsrest.web.controllers

import com.typesafe.scalalogging.LazyLogging
import edu.ucdavis.fiehnlab.ctsrest.client.types.FormulaResponse
import edu.ucdavis.fiehnlab.ctsrest.web.{CtsProxy, CtsProxyConfig}
import edu.ucdavis.fiehnlab.wcmc.utilities.casetojson.config.CaseClassToJSONSerializationAutoConfiguration
import org.junit.runner.RunWith
import org.scalatest.{Matchers, WordSpec}
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.web.servlet.ServletWebServerFactoryAutoConfiguration
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.boot.web.server.LocalServerPort
import org.springframework.test.context.TestContextManager
import org.springframework.test.context.junit4.SpringRunner

import scala.collection.JavaConverters._

/**
  * Created by diego on 2/15/2017.
  */
@RunWith(classOf[SpringRunner])
@SpringBootTest(classes = Array(classOf[CtsProxy]), webEnvironment = WebEnvironment.RANDOM_PORT)
class CtsControllerTest extends WordSpec with Matchers with LazyLogging {

  @Autowired
  val restTemplate: TestRestTemplate = null

  @LocalServerPort
  val port: Int = 0

  new TestContextManager(this.getClass).prepareTestInstance(this)

  val baseUrl = s"http://localhost:${port}/rest"

  "CtsController" should {

    "fromValues" in {
      val response = restTemplate.getForObject[Seq[String]](baseUrl + "/fromValues", classOf[Seq[String]])
      response should not be empty
      response.size should be > 10
      response should contain allOf("Chemical Name", "Human Metabolome Database", "InChIKey", "KEGG")
    }

    "toValues" in {
      val response = restTemplate.getForObject[Seq[String]](baseUrl + "/toValues", classOf[Seq[String]])
      response should not be empty
      response.size should be > 10
      response should contain allOf("Chemical Name", "Human Metabolome Database", "InChI Code", "InChIKey", "KEGG")
    }

    "getFormulaExpansion" in {
      val response = restTemplate.getForObject(baseUrl + "/expandformula/C2H6O", classOf[FormulaResponse])
      response shouldEqual FormulaResponse("C2H6O", "CCHHHHHHO", null)
    }

    "getSimpleConversion" in {
      val response = restTemplate.getForObject[Seq[Map[String, Any]]](baseUrl + "/convert/Chemical Name/InChIKey/alanine",
        classOf[Seq[Map[String, Any]]])
      println(response)

      response should not be empty
      response.head.getOrElse("fromIdentifier", "") should be ("Chemical Name")
      response.head.getOrElse("toIdentifier", "") should be ("InChIKey")
      response.head.getOrElse("searchTerm", "") should be ("alanine")
      response.head.getOrElse("results", Seq.empty) shouldBe a [Seq[_]]

      response.head.getOrElse("results", Seq.empty) should contain
          allOf("QNAYBMKLOCPYGJ-REOHCLBHSA-N", "QNAYBMKLOCPYGJ-UHFFFAOYSA-N", "QNAYBMKLOCPYGJ-UWTATZPHSA-N")
    }

    "testSecondCallHitsCache" ignore {
      // no cache header set, so test is ignored for now
      val response = restTemplate.getForEntity[Seq[String]](baseUrl + "/fromValues", classOf[Seq[String]])
      println(response.getHeaders.asScala.mkString("\n"))
      response.getHeaders.get("X-Proxy-Cache") shouldBe "MISS"

      val response2 = restTemplate.getForEntity[Seq[String]](baseUrl + "/fromValues", classOf[Seq[String]])
      response.getHeaders.get("X-Proxy-Cache") shouldBe "HIT"
    }
  }
}
